src/lib/operators/join_index.cpp

```diff
@@ -0,0 +1,346 @@
+                                     _predicate_condition);
+}
+
+std::shared_ptr<const Table> JoinIndex::_on_execute() {
+  _create_table_structure();
+
+  _perform_join();
+
+  return _output_table;
+}
+
+void JoinIndex::_create_table_structure() {
+  _output_table = std::make_shared<Table>();
+
+  _left_in_table = _input_left->get_output();
+  _right_in_table = _input_right->get_output();
+
+  _left_column_id = _column_ids.first;
+  _right_column_id = _column_ids.second;
+
+  const bool left_may_produce_null = (_mode == JoinMode::Right || _mode == JoinMode::Outer);
+  const bool right_may_produce_null = (_mode == JoinMode::Left || _mode == JoinMode::Outer);
+
+  // Preparing output table by adding columns from left table
+  for (ColumnID column_id{0}; column_id < _left_in_table->column_count(); ++column_id) {
+    auto nullable = (left_may_produce_null || _left_in_table->column_is_nullable(column_id));
+    _output_table->add_column_definition(_left_in_table->column_name(column_id), _left_in_table->column_type(column_id),
+                                         nullable);
+  }
+
+  // Preparing output table by adding columns from right table
+  for (ColumnID column_id{0}; column_id < _right_in_table->column_count(); ++column_id) {
+    auto nullable = (right_may_produce_null || _right_in_table->column_is_nullable(column_id));
+    _output_table->add_column_definition(_right_in_table->column_name(column_id),
+                                         _right_in_table->column_type(column_id), nullable);
+  }
+}
+
+void JoinIndex::_perform_join() {
+  DebugAssert(_mode != JoinMode::Cross, "Cross join mode is not supported by index join");
+
+  _right_matches.resize(_right_in_table->chunk_count());
+  _left_matches.resize(_left_in_table->chunk_count());
+
+  auto left_data_type = _left_in_table->column_type(_left_column_id);
+  auto right_data_type = _right_in_table->column_type(_right_column_id);
+
+  _pos_list_left = std::make_shared<PosList>();
+  _pos_list_right = std::make_shared<PosList>();
+
+  for (ChunkID chunk_id_left = ChunkID{0}; chunk_id_left < _left_in_table->chunk_count(); ++chunk_id_left) {
+    // initialize the data structures for left matches
+    _left_matches[chunk_id_left].resize(_left_in_table->get_chunk(chunk_id_left)->size());
+  }
+
+  // Scan all chunks for right input
+  for (ChunkID chunk_id_right = ChunkID{0}; chunk_id_right < _right_in_table->chunk_count(); ++chunk_id_right) {
+    const auto chunk_right = _right_in_table->get_chunk(chunk_id_right);
+    auto column_right = chunk_right->get_column(_right_column_id);
+    const auto indices = chunk_right->get_indices(std::vector<ColumnID>{_right_column_id});
+    _right_matches[chunk_id_right].resize(chunk_right->size());
+
+    std::shared_ptr<BaseIndex> index = nullptr;
+
+    if (indices.size() > 0) {
+      // We assume the first index to be efficient for our join
+      // as we do not want to spend time on evaluating the best index inside of this join loop
```

Comment:
This would mean a change of the interface from the other Joins. We were conservative in our approach and decided for this. If the interface breakage is allowed then an explicitly passed ColumnIndexType is indeed preferable.

Reply:
We should discuss this in person. The reason for not passing the index is that the interface is too complicated. We can have an arbitrary number of combinations here for many many chunks.
Also, our joining operators are currently not meant to be applied to a few chunks, we always join the whole table. Thus, the index join is very different from the index scan.

In the long run we probably want to have a function as part of the optimizer/statistics module which allows to request the optimal index for a particular chunk/access pattern.

Reply:
Okay then I understand why we need the nested-loop fallback here. Still, it’s a lot of duplicated code and considering that the nested-loop join is the operator with the longest compile time (6 min. in release mode), duplicating the most compile-time intensive part doesn’t seem to be a good solution.

Reply:
Outch ... 6 minutes is a lot. Good point.
@ Team Join: can you estimate the effort to extract the doubled code?

Reply:
Probably a day of work. We will get to that next week, we most likely won't make it before

Reply:
For most non-equi cases I can come up with right now, separating the operation into multiple joins should be relatively expensive. Imagine a outer-join where we need to keep track off all un-matched tuples. It's probably doable for all join methods, but I am not sure the effort is worth it.

Reply:
Gotcha